stdlib/public/core/Diffing.swift

```diff
@@ -0,0 +1,782 @@
+///
+/// - Note: `CollectionChanges` holds a reference to state used to run the
+///     difference algorithm, which can be exponentially larger than the
+///     changes themselves.
+fileprivate struct CollectionChanges<
+  SourceIndex : Comparable, TargetIndex : Comparable
+> {
+  typealias Endpoint = (x: SourceIndex, y: TargetIndex)
+
+  /// An encoding of change elements as an array of index pairs stored in
+  /// `pathStorage[pathStartIndex...]`.
+  ///
+  /// This encoding allows the same storage to be used to run the difference
+  /// algorithm, report the result, and repeat in place using
+  /// `formChanges`.
+  ///
+  /// The collection of changes between XABCD and XYCD is:
+  ///
+  ///   [.match(0..<1, 0..<1), .remove(1..<3), .insert(1..<2),
+  ///    .match(3..<5, 2..<4)]
+  ///
+  /// Which gets encoded as:
+  ///
+  ///   [(0, 0), (1, 1), (3, 1), (3, 2), (5, 4)]
+  ///
+  /// You can visualize it as a two-dimensional path composed of remove
+  /// (horizontal), insert (vertical), and match (diagonal) segments:
+  ///
+  ///     X A B C D
+  ///   X \ _ _
+  ///   Y    |
+  ///   C     \
+  ///   D     \
+  ///
+  private var pathStorage: [Endpoint]
+
+  /// The index in `pathStorage` of the first segment in the difference path.
+  private var pathStartIndex: Int
+
+  /// Creates a collection of changes from a difference path.
+  init(
+    pathStorage: [Endpoint], pathStartIndex: Int
+  ) {
+    self.pathStorage = pathStorage
+    self.pathStartIndex = pathStartIndex
+  }
+
+  /// Creates an empty collection of changes, i.e. the changes between two
+  /// empty collections.
+  private init() {
+    self.pathStorage = []
+    self.pathStartIndex = 0
+  }
+}
+
+extension CollectionChanges {
+  /// A range of elements removed from the source, inserted in the target, or
+  /// that the source and target have in common.
+  enum Element {
+    case removed(Range<SourceIndex>)
+    case inserted(Range<TargetIndex>)
+    case matched(Range<SourceIndex>, Range<TargetIndex>)
+  }
+}
+
+extension CollectionChanges : RandomAccessCollection {
+  typealias Index = Int
+
+  var startIndex: Index {
+    return 0
+  }
+
+  var endIndex: Index {
+    return Swift.max(0, pathStorage.endIndex - pathStartIndex - 1)
+  }
+
+  func index(after i: Index) -> Index {
+    return i + 1
+  }
+
+  subscript(position: Index) -> Element {
+    precondition((startIndex..<endIndex).contains(position))
+
+    let current = pathStorage[position + pathStartIndex]
+    let next = pathStorage[position + pathStartIndex + 1]
+
+    if current.x != next.x && current.y != next.y {
+      return .matched(current.x..<next.x, current.y..<next.y)
+    } else if current.x != next.x {
+      return .removed(current.x..<next.x)
+    } else { // current.y != next.y
+      return .inserted(current.y..<next.y)
+    }
+  }
+}
+
+extension CollectionChanges : CustomStringConvertible {
+  var description: String {
+    return String(describing: Array(self))
```

Comment:
We have a dedicated function for this in the stdlib:

```suggestion
    return _makeCollectionDescription()
```